pyro/poutine/messenger.py

```diff
@@ -114,13 +114,37 @@ def _process_message(self, msg):
         Process the message by calling appropriate method of itself based
         on message type. The message is updated in place.
         """
-        return getattr(self, "_pyro_{}".format(msg["type"]))(msg)
+        if hasattr(self, "_pyro_{}".format(msg["type"])):
+            return getattr(self, "_pyro_{}".format(msg["type"]))(msg)
+        return None
 
     def _postprocess_message(self, msg):
+        if hasattr(self, "_pyro_post_{}".format(msg["type"])):
+            return getattr(self, "_pyro_post_{}".format(msg["type"]))(msg)
         return None
 
-    def _pyro_sample(self, msg):
-        return None
+    @classmethod
+    def register(cls, fn=None, type=None, post=None):
+        """
+        @SomeMessengerClass.register
+        def some_function(...)
+        """
+        if fn is None:
+            return lambda x: cls.register(x, type=type, post=post)
 
-    def _pyro_param(self, msg):
-        return None
+        if type is None:
+            type = fn.__code__.co_name
```

Comment:
What is the difference between `fn.__name__` and `fn.__code__.co_name`? Should I start avoiding `fn.__name__`?

Reply:
I think `fn.__code__.co_name` will refer to the name in the original code object, whereas `__name__` can refer to an intermediate variable name?  It's probably not a great idea to provide a default `type` anyway, I should just remove that.